### PR TITLE
Coverity fixes plus fix for potential multithreading issue

### PR DESCRIPTION
**Pull request by jclarkeSTFC**
Most fixes were for copying in instead of moving in to a method that I've recently added to fix other Coverity warnings about null objects (#38100). The changes in `FindPeaksConvolve` were because `m_pdf` was being either read or written to without the lock, so a potential bug.

### To test:

Try running `FindPeaksConvolve`, e..g the test code in #38135.

*This does not require release notes* because **it has no impact on the user.**

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp [hunk 1]
```diff
@@ -379,4 +361,28 @@ ConvertFitFunctionForMuonTFAsymmetry::getTFAsymmFitFunction(const Mantid::API::I
...
+  inBrace->addFunction(constant);
+  inBrace->addFunction(userFunc);
+  auto norm = FunctionFactory::Instance().createInitialized("name = FlatBackground, A0 "
+                                                            "=" +
+                                                            std::to_string(normValue));
+  auto product = Kernel::DynamicPointerCastHelper::dynamicPointerCastWithCheck<CompositeFunction, API::IFunction>(
+      FunctionFactory::Instance().createFunction("ProductFunction"));
+  product->addFunction(norm);
+  product->addFunction(inBrace);
+  auto composite = Kernel::DynamicPointerCastHelper::dynamicPointerCastWithCheck<CompositeFunction, API::IFunction>(
```
Comment (RichardWaiteSTFC) [inline Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp:379]: Just a question to help me - why is this necessary?
Note this line is also present in `PawleyFunction.cpp` and a few other places
https://github.com/mantidproject/mantid/blob/005e651f12b64e9dc38d472d5d74627d51a99af7/Framework/CurveFitting/src/Functions/PawleyFunction.cpp#L458

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks good to me, followed instructions for `FindPeaksConvolve` in #36352 - just a quick question for my own understanding!

File: Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp (hunk 1 above)
Comment (jclarkeSTFC) [reply Framework/Muon/src/ConvertFitFunctionForMuonTFAsymmetry.cpp:379]: If we do `dynamic_pointer_cast<T>` and then use the result without checking for null, then Coverity complains. In #38100 I added `dynamicPointerCastWithCheck` to include the null check before returning the pointer, (and save duplicating `if ( thing == nullptr)` in several places.
